File: search-a-2d-matrix-ii.py
# // Time Complexity : O(M+N)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : YES


# // Three line explanation of solution in plain english
# We try to employ 2 pointer approach and we try to eliminate either a row or a column by comparing the pivot element (Top-right)
# If the pivot is greater than the target then we ignore the top row (Move down - change pivot to down)
# If the pivot is less than the target then we ignore the right row (Move left - change the pivot to left)
#Until we get a single element - if it is the target then True - Else False


# // Your code here along with comments explaining your approach
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])
        nr, nc = 0, n-1
        while min(nr, nc) >= 0 and nr < m and nc < n:
            currVal = matrix[nr][nc]
            if currVal == target:
                return True
            elif target <= currVal:
                nr, nc = nr, nc-1
            else:
                nr, nc = nr+1, nc
        return False

File: Searcha2DMatrixII.cs

// Time Complexity :O(N)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :Yes
// Three line explanation of solution in plain english

// Your code here along with comments explaining your approach

// This code searches for a target value in a sorted 2D matrix.
// It starts from the top-right corner and iterates either left or down based on comparisons with the target.
// If the target is found, it returns true; otherwise, it returns false after traversing the matrix.

public class Solution {
    public bool SearchMatrix(int[][] matrix, int target) {
        if(matrix == null || matrix.Length ==0) return false;
        int m = matrix.Length;
        int n = matrix[0].Length;

        int i=0;
        int j = n-1;

        while(i < m && j>=0)
        {
            if(matrix[i][j] == target)
            {
                return true;
            }
            else if(matrix[i][j] > target)
            {
                j--; // move left
            }
            else
            {
                i++; //down
            }
        }
        return false;
    }
}